mesonbuild/msetup.py

```diff
@@ -308,7 +312,8 @@ def _generate(self,
                     os.unlink(cdf)
             raise
 
-        return captured_compile_args
+        # Even with the asserts, mypy's type checking thinks this could be 'Any' type, not the None|dict we know it to be, hence the suppression comment -
+        return captured_compile_args # type: ignore[no-any-return]
```

Comment:
If you're suppressing errors there's a 99% chance you misunderstood how things work. In the remaining 1% case you should have a better comment than "I thought this was supposed to work but it doesn't so just ignore it" to explain why.

Reply:
That's intriguing. Supposing we follow through your 1% case scenario, where there actually is no misunderstanding and we actually have given mypy everything needed to deduce the type that it's getting wrong:  You say you don't like a comment that explains why we think mypy has enough to deduce a type but which it fails to deduce.  So what are you proposing would be a better kind of comment in this scenario?

Reply:
The problem is "but which it fails to deduce".

If you've given it everything it needs to deduce the type, but it isn't deducing it, then either the code isn't giving it the information you think it is, or else the mypy documentation should discuss the caveat and serve as a reference, or else it's a bug in mypy and deserves to be reported.

(Either way using a cast is often better.)

Reply:
Well, this statement would be a lot easier to swallow if even just 1 of the currently 30 instances of `# type: ignore` that are scattered throughout meson had any hint of an accompanying explanation and/or apology, never even mind their total absence of any reference to a mypy bug that, according to you, must be used if one has to resort to suppressing a mypy type warning for something that otherwise definitely does provide enough type info, that leaves concluding a mypy bug as the only remaining conclusion.

Reply:
Ha. Well, I've found 1 instance of this in mtest.py, ln 126.

Reply:
> This is very clever. `__init__` is not a constructor, it's an initializer, therefore it's safe to call more than once. [...] Mypy doesn't like this, because it's so clever.

> ignore all lints for this file, since it is run by python2 as well

> Mypy thinks these overlap, but since List is invariant they don't, `List[str]` is not a valid input to `List[str | BuildTarget]`. pylance/pyright gets this right, but for mypy we have to ignore the error

> Mypy doesn't follow hasattr, and it's pretty easy to visually inspect that this is correct, so we'll just ignore it.

> windll only exists on windows, so mypy will get mad

> We are doing the equivalent to setattr here and mypy does not like it

> `# type: ignore # because the entire point of this is to diverge`

> This trickery with type() gets us the class of the linker so we can initialize a new copy for the Rust Compiler TODO rewrite this without type: ignore

> Mypy can't figure out that token is overridden inside the while loop

> Pylance can figure out that these do not, in fact, overlap, but mypy can't

I'll admit that not all cases do have documentation. The gnome.py and pkgconfig.py issue is that `['key1', 'key2']` can't be used as indexes into a TypedDict without repeating the *entire* list as T.Literal. The linkers.py issue is that a mixin doesn't have access to its base class. The type_checking.py issue is that a KwargInfo is generic according to types=, but a convertor can reduce that further and this cannot be captured with type annotations. The mlog.py issue is that attr-defined on a stdlib type is being ignored to smuggle our own value persistently on the object it affects.

Reply:
In this case it seems the solution was to annotate the original value as not just None, but `T.Optional[TheOtherThingItCouldBe]`, so probably not a mypy bug. This is the problem with being quick to mark a test as "expected to fail", especially in *newly added* code rather than when adding tests after the fact.

e.g. arglist.py
> TODO: fix 'Invalid index type' and 'Incompatible types in assignment' erros

which was added in a PR that exclusively added typing to existing code, and inherently has more lax standards; in the generic case, when adding tests to existing code, it's okay to mark them as expectedfail "but we do not know why".

Reply:
See commits 1 and 3 here: #11938

I added comments to 3 cases, and removed 3 more entirely because it turns out we *can* check the mixins.

Reply:
Many of those mypy work-around excuse comments you quote look very much like this scenario, by which I mean they certainly don't reference a mypy bug that you eventually said is what should be used to accompany these suppressions.

This -
> Mypy doesn't follow hasattr, and it's pretty easy to visually inspect that this is correct, so we'll just ignore it.

describes a very similar scenario to what's been going on here.  In my case, mypy doesn't follow some pretty clear to see type narrowing deduction in what is (until anyone suggests otherwise ... i.e. I'm still waiting; are you still insisting it's not a real assert?), valid and clear python.

My impression is that it's actually quite misleading and unhelpful to say "there's a 99% chance you misunderstood how things work".  Seems like it's more a case of:  mypy's type checking can simply be wrong, but with some persistence and wrangling/restructuring - perhaps even contortions 😄-  of the code, it might be possible to get mypy to figure things out.

In future, and especially if it can be explained in a sentence or two, if you have insight on what you think is a misunderstanding, there's a 100% chance that it'll be more useful and productive to share that insight, rather than leaving nothing more than a, "That's not a real assert" kind of remark.  This whole exchange could have been far simpler if you'd have just said something like -

"It looks like you're not giving enough type info to mypy.  Here's why I think this: ..."

or -

"Use of '# type: ignore' should be a last resort if you think you've given mypy enough to go on, [which it looks like you haven't because ...] / [which it looks like you have but I can't suggest what more you could do to help it out].  Any last-resort suppressions should be accompanied with a good explanation and preferably (but not actually required, as evidence by the many existing unaccompanied uses) a comment pointing to a mypy bug".

Reply:
And by the way, if it's valid code from which the reader can trivially deduce type-correctness, then I'd say it very much is a mypy bug.  Call it a limitation if you prefer, but a work-around from valid, type-correct code is still a work-around.

Reply:
> describes a very similar scenario to what's been going on here. In my case, mypy doesn't follow some pretty clear to see type narrowing deduction

The difference is that not everything is, in fact, a type narrowing deduction:
- mypy literally does not implement hasattr and is expected to get it wrong for the reason in the comment
- whereas it is "supposed" to type narrow on isinstance / x is None and was simply failing to do so, here, and the comment is saying (paraphrased):
  > this code is correct and mypy supports it and correctly analyzes it. Except actually it does not and I'm not sure why


> preferably (but not actually required, as evidence by the many existing unaccompanied uses) a comment pointing to a mypy bug

But also I didn't insist on an url comment, I merely said that there should exist references. I had intended to mean, by that, that it should be possible to read the mypy documentation / bugtracker and conclude some authoritative fact like "mypy does not support hasattr, so we don't even try to type narrow" or "this is definitely a mypy bug because pylance gets it correct, thereby proving the annotation is not faulty" or "this is a bug, here's an url link describing why". Apologies for the confusion.

Again, this is no different from e.g. a unittest. If a PR adds code that causes a unittest to fail, and you want to mark the unittest as "expectedfail", I expect *some form* of explanation for why the unittest failure isn't a problem, and that explanation cannot be "I believe I did everything right, the test still fails, it must be a problem in the test software". But "I definitely 100% can verify that the test software is buggy because it doesn't know how to frobnicate the wolpertinger and this test does that" would be a good explanation.


If in doubt, I would prefer that rather than "use type: ignore as a last resort" / "mark this test as expectedfail" a self-review comment is made saying "this test fails and I'm not sure why, any ideas?"